src/audio/kpb.c

```diff
@@ -0,0 +1,541 @@
+					  MAX_BUFFER_SIZE - LPSRAM_SIZE);
+	cd->his_buf_lp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_LP,
+					  LPSRAM_SIZE);
+
+	if (cd->his_buf_hp.sta_addr || cd->his_buf_lp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+		return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr +
+	(MAX_BUFFER_SIZE - LPSRAM_SIZE);
+
+	cd->his_buf_lp.end_addr = cd->his_buf_lp.sta_addr +
+	LPSRAM_SIZE;
+
+#elif KPB_NO_OF_HISTORY_BUFFERS == 1
+
+	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+					  MAX_BUFFER_SIZE);
+
+	if (cd->his_buf_hp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+	return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr + MAX_BUFFER_SIZE;
+
+#else
+#error "Wrong number of key phrase buffers configured"
+#endif
+
+	/* TODO: zeroes both buffers */
+
+	/* Initialize clients data */
+	for (i = 0; i < MAX_NO_OF_CLIENTS; i++)
+		cd->clients[i].state = KPB_CLIENT_UNREGISTERED;
+
+	return ret;
+}
+
+/**
+ * kpb_copy() - copy real time input stream into sink buffer,
+ * and in the same time buffers that input for
+ * later use by some of clients.
+ *
+ * @arg1:  kpb component.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_copy(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	struct comp_buffer *sink;
+
+	trace_kpb("kpb_copy()");
+
+	/* Get source and sink buffers */
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 source_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       sink_list);
+	/* Process source data */
+	/* Check if source have data and valid pointer to read from */
+	if (source->avail > 0 && source && sink && source->r_ptr) {
+		/* Real time copying */
+		if (sink->free >= source->avail && sink->w_ptr) {
+			/* Sink and source are both ready and have space */
+			memcpy(sink->w_ptr, source->r_ptr,
+			       source->avail);
+
+			/* update source & snik data*/
+			comp_update_buffer_produce(sink, source->avail);
+			comp_update_buffer_consume(source, source->avail);
+		} else {
+			;/* What should we do if sink isn't ready(state/size)?
+			  * Simple skip is OK?
+			  */
+		}
+		/* Buffer source data internally for future use by clients */
+		if (source->avail <= MAX_BUFFER_SIZE)
+			kpb_buffer_data(kpb, source);
+	} else
+		ret = -EINVAL;
+	return ret;
+}
+
+/**
+ * kpb_buffer_data() - buffer real time data stream in
+ * the internal buffer.
+ *
+ * @arg1:  kpb component.
+ * @arg2:  pointer to the source.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source)
+{
+trace_kpb("kpb_buffer_data()");
+
+#if KPB_NO_OF_HISTORY_BUFFERS == 2
+	int size_to_copy = source->avail;
```

Comment:
Variables should be declared at the beginning of function.

Reply:
Is KPB_NO_OF_HISTORY_BUFFERS a config option?

Reply:
@dbaluta, that's the plan, yes.